### ISSUE TITLE
LightningDataModule: GPU data augmentation support

### ISSUE THREAD

**Issue by adamjstewart** [PyTorchLightning/pytorch-lightning#10469]
## 🚀 Feature

Data augmentation libraries like Kornia support computation directly on the GPU, greatly speeding up the rate at which images can be sampled. I would like to be able to perform these kinds of GPU transforms in a LightningDataModule.

### Motivation

In [TorchGeo](https://github.com/microsoft/torchgeo), we use PyTorch Lightning to organize reproducible benchmarks for geospatial datasets. Currently, we have a set of LightningDataModules for each dataset and a much smaller number of LightningModules for each task (semantic segmentation, classification, regression, etc.). However, the LightningDataModule doesn't seem to know anything about the GPU, and the LightningModule doesn't seem to know anything about the LightningDataModule. Because of this, if we want to perform dataset-specific augmentations on the GPU, we're forced to create a separate LightningModule for each dataset, increasing code duplication and defeating the whole purpose of PyTorch Lightning.

### Pitch

The purpose of a LightningDataModule is to handle all dataset-specific loading and augmentation so that a generic LightningModule can handle the actual training and evaluation. However, in order to take advantage of GPU-accelerated libraries like Kornia, we're currently forced to move this logic to a LightningModule. As datasets continue to increase in size, direct support for GPU-accelerated transforms in LightningDataModules will increase in importance.

### Alternatives

So far the only alternative we've found is to create a different LightningModule for each dataset and include the data augmentation there. If there's a better alternative to this we would love to know about it!

@calebrob6 @isaaccorley

**Comment by tchaton** [comment]
Hey @adamjstewart,

Awesome work on TorchGeo, you can actually easily enable GPU transform on your DataModule.

Check this DataModule in Lightning Flash: https://github.com/PyTorchLightning/lightning-flash/blob/master/flash/core/data/new_data_module.py#L240.

The DataModule implements an `on_after_batch_transfer` hook which will be attached to the Model during training.

In Flash, the dataset owns their transform, and we extract the `on_after_batch_transfer_fn` directly from the InputTransform.

Have a look at our tutorial: https://github.com/PyTorchLightning/lightning-flash/blob/master/flash_examples/flash_components/custom_data_loading.py

Best,
T.C

**Comment by adamjstewart** [comment]
@tchaton the docs for [on_after_batch_transfer](https://pytorch-lightning.readthedocs.io/en/latest/api/pytorch_lightning.core.hooks.html#pytorch_lightning.core.hooks.DataHooks.on_after_batch_transfer) mention using `self.trainer.training` but I can't see where this attribute is defined. Is this a boolean? Also, one of the links that you shared before seems to be dead now.

**Comment by tchaton** [comment]
Hey @adamjstewart,

Yes, it is a boolean. It is defined there on the Trainer: https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/trainer/trainer.py#L2028.

Best,
T.C
